include/seastar/core/sstring.hh

```diff
@@ -346,6 +346,29 @@ public:
         return *this;
     }
 
+    /**
+     * Append the given character to the end of the string
+     * @param ch  The character to append
+     * @note unlike @c std::basic_string, sstring does not preallocate, so
+     *       this call always lead to reallocation if external storage is
+     *       used.
+     */
+    constexpr void push_back(char_type ch) {
+        if (is_internal()) {
+            if (static_cast<Size>(u.internal.size) < max_size) {
+                u.internal.str[u.internal.size++] = ch;
+                if (NulTerminate) {
+                    u.internal.str[u.internal.size] = '\0';
+                }
+                return;
+            }
+        }
+        basic_sstring new_str(initialized_later(), u.external.size + 1);
+        std::copy(begin(), end(), new_str.begin());
```

Comment:
The documentation https://en.cppreference.com/w/cpp/string/basic_string/push_back says that push_back() should have "amortized constant" complexity. It means you cannot do this copy on every push_back() call - it would make it O(n) complexity. I think you need to multiply the size of the string by something (1.1 or 2 or whatever) and until you reach the allocated size, just copy one character and not the entire string.

Reply:
We don't have unused capacity, so can't do that.

I think push_back is bad since it can't be implemented efficiently and will lead to accidentally quadratic users.

Reply:
If we can't do that, we shouldn't implement this function. Its complexity is part of its API.

Reply:
how about trading the implementation for a comment, noting that we don't intend to implement this because the performance issues due to the inherent design decision made by `sstring`?

Reply:
Nobody will read a comment on push_back. Everyone already knows how it works.

Reply:
@avikivity hi Avi, i didn't mean to add a comment near by this function. i wanted to put a comment **in the place of it** explaining why we don't have it. to explain that it is a design decision.